This patch gets a failure on TestLogic for explain_debug:
```
--- FAIL: TestLogic (3.27s)
        logic_test.go:698: testdata/explain_debug:110: expected:
                "0"    "0"    "(1,"  "2,"   
                "3)"   "ROW"  "1"    "1"    
                "(4,"  "5,"   "6)"   "ROW"  
                but found:
                "0"    "0"    "(1,"  "2,"   
                "3)"   "ROW"  "1"    "1"    
                "(4,"  "5,"   "6)"   "ROW"  
                "2"    "2"    "(1,"  "2,"   
                "3)"   "ROW"  "3"    "3"    
                "(4,"  "5,"   "6)"   "ROW"  
```

Seems EXPLAIN(DEBUG) is now seeing the value rows two times. How come?

cc @RaduBerinde

--- a/sql/explain.go
+++ b/sql/explain.go
@@ -76,20 +76,14 @@ func (p *planner) Explain(n *parser.Explain, autoCommit bool) (planNode, *roachp
 			return nil, roachpb.NewError(err)
 		}
 		p.txn.Context = opentracing.ContextWithSpan(p.txn.Context, sp)
-
 	}
 
 	plan, err := p.makePlan(n.Statement, autoCommit)
 	if err != nil {
 		return nil, err
 	}
-	if err := plan.Start(); err != nil {
-		return nil, err
-	}
 	switch mode {
 	case explainDebug:
-		plan.MarkDebug(mode)
-
 		// Wrap the plan in an explainDebugNode.
 		return &explainDebugNode{plan}, nil
 
@@ -104,7 +98,6 @@ func (p *planner) Explain(n *parser.Explain, autoCommit bool) (planNode, *roachp
 		return v, nil
 
 	case explainTrace:
-		plan.MarkDebug(explainDebug)
 		return (&sortNode{
 			ordering: []columnOrderInfo{{len(traceColumns), encoding.Ascending}, {2, encoding.Ascending}},
 			columns:  traceColumns,
@@ -217,8 +210,14 @@ var debugColumns = []ResultColumn{
 func (*explainDebugNode) Columns() []ResultColumn { return debugColumns }
 func (*explainDebugNode) Ordering() orderingInfo  { return orderingInfo{} }
 
-func (n *explainDebugNode) PErr() *roachpb.Error  { return n.plan.PErr() }
-func (n *explainDebugNode) Start() *roachpb.Error { return n.plan.Start() }
+func (n *explainDebugNode) PErr() *roachpb.Error { return n.plan.PErr() }
+func (n *explainDebugNode) Start() *roachpb.Error {
+	if pErr := n.plan.Start(); pErr != nil {
+		return pErr
+	}
+	n.plan.MarkDebug(explainDebug)
+	return nil
+}
 
 func (n *explainDebugNode) Next() bool { return n.plan.Next() }
 

--- a/sql/pgwire_test.go
+++ b/sql/pgwire_test.go
@@ -412,6 +412,9 @@ func TestPGPreparedQuery(t *testing.T) {
 			base.Params(1).Results(1),
 			base.Params(nil).Results(1),
 		},
+		"INSERT INTO d.T VALUES ($1::INT) RETURNING 1": {
+			base.Params(1).Results(1),
+		},
 		"INSERT INTO d.T VALUES ($1) RETURNING $1": {
 			base.Params(1).Results(1),
 			base.Params(3).Results(3),

--- a/sql/plan.go
+++ b/sql/plan.go
@@ -456,7 +456,9 @@ type planNode interface {
 	SetLimitHint(numRows int64, soft bool)
 
 	// MarkDebug puts the node in a special debugging mode, which allows
-	// DebugValues to be used.
+	// DebugValues to be used. This should be called after Start() and
+	// before the first call to Next() since it may need to recurse into
+	// sub-nodes created by Start().
 	MarkDebug(mode explainMode)
 }
 

--- a/sql/subquery.go
+++ b/sql/subquery.go
@@ -69,6 +69,10 @@ func (v *subqueryVisitor) VisitPre(expr parser.Expr) (recurse bool, newExpr pars
 		return false, expr
 	}
 
+	if v.pErr = plan.Start(); v.pErr != nil {
+		return false, expr
+	}
+
 	if exists != nil {
 		// For EXISTS expressions, all we want to know is if there is at least one
 		// result.

--- a/sql/testdata/explain_plan
+++ b/sql/testdata/explain_plan
@@ -31,13 +31,13 @@ query ITT colnames
 EXPLAIN VALUES (1, 2, 3), (4, 5, 6)
 ----
 Level  Type    Description
-0      values  3 columns, 2 rows
+0      values  3 columns
 
 query ITT colnames
 EXPLAIN VALUES (1)
 ----
 Level  Type    Description
-0      values  1 column, 1 row
+0      values  1 column
 
 query ITT colnames
 EXPLAIN SELECT * FROM t LIMIT 1 OFFSET 1

--- a/sql/testdata/subquery
+++ b/sql/testdata/subquery
@@ -258,7 +258,7 @@ EXPLAIN SELECT * FROM (SELECT * FROM (VALUES (1, 8, 8), (3, 1, 1), (2, 4, 4)) AS
 ----
 0 sort   +foo1
 1 sort   +moo2
-2 values 3 columns, 3 rows
+2 values 3 columns
 
 query II colnames
 SELECT a, b FROM (VALUES (1, 2, 3), (3, 4, 7), (5, 6, 10)) AS foo (a, b, c) WHERE a + b = c

--- a/sql/trace.go
+++ b/sql/trace.go
@@ -52,7 +52,11 @@ func (*explainTraceNode) Ordering() orderingInfo  { return orderingInfo{} }
 
 func (n *explainTraceNode) PErr() *roachpb.Error { return nil }
 func (n *explainTraceNode) Start() *roachpb.Error {
-	return n.plan.Start()
+	if pErr := n.plan.Start(); pErr != nil {
+		return pErr
+	}
+	n.plan.MarkDebug(explainDebug)
+	return nil
 }
 func (n *explainTraceNode) Next() bool {
 	first := n.rows == nil

--- a/sql/values.go
+++ b/sql/values.go
@@ -27,17 +27,29 @@ import (
 	"github.com/cockroachdb/cockroach/util/encoding"
 )
 
+type valuesNode struct {
+	n        *parser.ValuesClause
+	p        *planner
+	columns  []ResultColumn
+	ordering columnOrdering
+	rows     []parser.DTuple
+
+	nextRow       int           // The index of the next row.
+	invertSorting bool          // Inverts the sorting predicate.
+	tmpValues     parser.DTuple // Used to store temporary values.
+}
+
 // ValuesClause constructs a valuesNode from a VALUES expression.
 func (p *planner) ValuesClause(n *parser.ValuesClause) (planNode, *roachpb.Error) {
 	v := &valuesNode{
-		rows: make([]parser.DTuple, 0, len(n.Tuples)),
+		p: p,
+		n: n,
 	}
 	if len(n.Tuples) == 0 {
 		return v, nil
 	}
 
 	numCols := len(n.Tuples[0].Exprs)
-	rowBuf := make(parser.DTuple, len(n.Tuples)*numCols)
 	v.columns = make([]ResultColumn, 0, numCols)
 
 	for num, tuple := range n.Tuples {
@@ -45,21 +57,18 @@ func (p *planner) ValuesClause(n *parser.ValuesClause) (planNode, *roachpb.Error
 			return nil, roachpb.NewUErrorf("VALUES lists must all be the same length, %d for %d", a, e)
 		}
 
-		// Chop off prefix of rowBuf and limit its capacity.
-		row := rowBuf[:numCols:numCols]
-		rowBuf = rowBuf[numCols:]
-
 		for i := range tuple.Exprs {
+			// TODO(knz): We need to expand subqueries two times, once here
+			// and once in Start() below, until the logic for select is split
+			// between makePlan and Start(). This is because a first call is
+			// needed for typechecking, and a separate call is needed to
+			// select indexes.
 			expr, pErr := p.expandSubqueries(tuple.Exprs[i], 1)
 			if pErr != nil {
 				return nil, pErr
 			}
 			var err error
 			typ, err := expr.TypeCheck(p.evalCtx.Args)
-			if err != nil {
-				return nil, roachpb.NewError(err)
-			}
-			expr, err = p.parser.NormalizeExpr(p.evalCtx, expr)
 			if err != nil {
 				return nil, roachpb.NewError(err)
 			}
@@ -70,25 +79,50 @@ func (p *planner) ValuesClause(n *parser.ValuesClause) (planNode, *roachpb.Error
 			} else if typ != parser.DNull && !typ.TypeEqual(v.columns[i].Typ) {
 				return nil, roachpb.NewUErrorf("VALUES list type mismatch, %s for %s", typ.Type(), v.columns[i].Typ.Type())
 			}
-			row[i], err = expr.Eval(p.evalCtx)
-			if err != nil {
-				return nil, roachpb.NewError(err)
-			}
 		}
-		v.rows = append(v.rows, row)
 	}
 
 	return v, nil
 }
 
-type valuesNode struct {
-	columns  []ResultColumn
-	ordering columnOrdering
-	rows     []parser.DTuple
+func (n *valuesNode) Start() *roachpb.Error {
+	if n.n == nil {
+		return nil
+	}
 
-	nextRow       int           // The index of the next row.
-	invertSorting bool          // Inverts the sorting predicate.
-	tmpValues     parser.DTuple // Used to store temporary values.
+	n.rows = make([]parser.DTuple, 0, len(n.n.Tuples))
+
+	// This node is coming from a SQL query (as opposed to sortNode and
+	// others that create a valuesNode internally for storing results
+	// from other planNodes), so it needs to evaluate expressions.
+
+	numCols := len(n.columns)
+	rowBuf := make(parser.DTuple, len(n.n.Tuples)*numCols)
+
+	for _, tuple := range n.n.Tuples {
+		// Chop off prefix of rowBuf and limit its capacity.
+		row := rowBuf[:numCols:numCols]
+		rowBuf = rowBuf[numCols:]
+
+		for i := range tuple.Exprs {
+			// TODO(knz): see comment above about expandSubqueries in ValuesClause().
+			expr, pErr := n.p.expandSubqueries(tuple.Exprs[i], 1)
+			if pErr != nil {
+				return pErr
+			}
+			var err error
+			expr, err = n.p.parser.NormalizeExpr(n.p.evalCtx, expr)
+			if err != nil {
+				return roachpb.NewError(err)
+			}
+			row[i], err = expr.Eval(n.p.evalCtx)
+			if err != nil {
+				return roachpb.NewError(err)
+			}
+		}
+		n.rows = append(n.rows, row)
+	}
+	return nil
 }
 
 func (n *valuesNode) Columns() []ResultColumn {
@@ -112,10 +146,6 @@ func (n *valuesNode) DebugValues() debugValues {
 		value:  n.rows[n.nextRow-1].String(),
 		output: debugValueRow,
 	}
-}
-
-func (n *valuesNode) Start() *roachpb.Error {
-	return nil
 }
 
 func (n *valuesNode) Next() bool {
@@ -228,9 +258,8 @@ func (n *valuesNode) ExplainPlan(_ bool) (name, description string, children []p
 		}
 		return "s"
 	}
-	description = fmt.Sprintf("%d column%s, %d row%s",
-		len(n.columns), pluralize(len(n.columns)),
-		len(n.rows), pluralize(len(n.rows)))
+	description = fmt.Sprintf("%d column%s",
+		len(n.columns), pluralize(len(n.columns)))
 	return name, description, nil
 }
 

--- a/sql/values_test.go
+++ b/sql/values_test.go
@@ -114,6 +114,10 @@ func TestValues(t *testing.T) {
 			t.Errorf("%d: error_expected=%t, but got error %v", i, tc.ok, pErr)
 		}
 		if plan != nil {
+			if pErr := plan.Start(); pErr != nil {
+				t.Errorf("%d: unexpected error in start: %v", i, pErr)
+				continue
+			}
 			var rows []parser.DTuple
 			for plan.Next() {
 				rows = append(rows, plan.Values())

